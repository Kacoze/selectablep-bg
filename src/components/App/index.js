import React, { Component } from 'react';
import fetch from 'isomorphic-fetch';
import styles from './style.scss';

const initialState = {
  suggestions: [],
};

class App extends Component {
  state = {
    ...initialState,
    background: '#000000',
    colors: [],
    inputValue: '',
  };

  componentDidMount() {
    fetch('http://www.mocky.io/v2/5a37a7403200000f10eb6a2d')
      .then(res => res.json())
      .then(colors => this.setState({ colors }))
      .catch(err => console.log(err));
  }

  handleInputValue = ({ target: { value } }) => {
    const { colors } = this.state;
    if (value.length > 1) {
      const regex = new RegExp(value);
      const suggestions = [];
      colors.forEach((i) => {
        if (regex.test(i.name)) {
          suggestions.push({ ...i, bold: i.name.replace(value, `<b>${value}</b>`) });
        }
      });
      this.setState({ suggestions, inputValue: value });
    } else {
      this.setState({ ...initialState });
    }
  }

  handleColorPick = (background) => {
    const input = this.colorInput;
    if (input) {
      input.value = background;
    }
    this.setState({ inputValue: background });
  }

  handleColorAccept = (color) => {
    const { colors, inputValue } = this.state;
    const picked = colors.filter(i => i.name === inputValue)[0];
    if (picked || color) {
      this.setState({ background: picked ? picked.hex : color });
    }
  }

  convertHex = (hex, opacity) => {
    const nhex = hex.replace('#', '');
    const r = parseInt(nhex.substring(0, 2), 16);
    const g = parseInt(nhex.substring(2, 4), 16);
    const b = parseInt(nhex.substring(4, 6), 16);

    return `rgba(${r},${g},${b},${opacity / 100})`;
  }

  render() {
    const { background, suggestions } = this.state;
    const bg = { background: this.convertHex(background, 50) };
    return (
      <div style={bg} className={styles.container}>
        <div className={styles.autoselect}>
          <div className={styles.inputContainer}>
            <input
              type="text"
              className={styles.autoselectInput}
              onChange={this.handleInputValue}
              // eslint-disable-next-line
              ref={i => this.colorInput = i}
            />
            <ul className={styles.options}>
              {suggestions.map(i => (
                <li key={`${i.name}${Math.random()}`}>
                  <button
                    className={styles.suggestion}
                    // eslint-disable-next-line
                    dangerouslySetInnerHTML={{ __html: i.bold }}
                    onClick={() => this.handleColorPick(i.name)}
                  />
                </li>
              ))}
            </ul>
          </div>
          <button className={styles.applyBtn} onClick={this.handleColorAccept}>Apply</button>
        </div>
      </div>
    );
  }
}

export default App;
