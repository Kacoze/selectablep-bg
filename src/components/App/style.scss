@import '~shared/styles/index.scss';

* {
  box-sizing: border-box;
  transition: all 1s;
}

.container {
  min-height: 100vh;
  width: 100vw;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
  align-items: center;
}
.autoselect {
  display: flex;
  position: relative;
  top: -100px;
  width: 500px;
  max-width: 100vw;
}

.applyBtn {
  height: 50px;
}

.autoselectInput {
  width: 100%;
  height: 50px;
  border-radius: 10px;
  font-size: 30px;
}

.inputContainer {
  width: 100%;
}

.suggestion {
  width: 100%;
  border: none;
}

.options {
  list-style: none;
  margin: 0;
  padding: 0;
}
