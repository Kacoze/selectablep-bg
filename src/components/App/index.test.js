import React from 'react';
import renderer from 'react-test-renderer';
import App from './index';

it('renders without crashing and match snapshoot', () => {
  const component = renderer.create(<App />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  component.unmount();
});

it('changes bg color', () => {
  const component = renderer.create(<App />);
  component.root.instance.handleColorAccept('#dddddd');
  expect(component.toJSON()).toMatchSnapshot();
  component.unmount();
});
