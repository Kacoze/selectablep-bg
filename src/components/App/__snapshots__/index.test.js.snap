// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes bg color 1`] = `
<div
  className="container"
  style={
    Object {
      "background": "rgba(221,221,221,0.5)",
    }
  }
>
  <div
    className="autoselect"
  >
    <div
      className="inputContainer"
    >
      <input
        className="autoselectInput"
        onChange={[Function]}
        type="text"
      />
      <ul
        className="options"
      />
    </div>
    <button
      className="applyBtn"
      onClick={[Function]}
    >
      Apply
    </button>
  </div>
</div>
`;

exports[`renders without crashing and match snapshoot 1`] = `
<div
  className="container"
  style={
    Object {
      "background": "rgba(0,0,0,0.5)",
    }
  }
>
  <div
    className="autoselect"
  >
    <div
      className="inputContainer"
    >
      <input
        className="autoselectInput"
        onChange={[Function]}
        type="text"
      />
      <ul
        className="options"
      />
    </div>
    <button
      className="applyBtn"
      onClick={[Function]}
    >
      Apply
    </button>
  </div>
</div>
`;
